Modules/_abc.c

```diff
@@ -0,0 +1,639 @@
+    return PyType_Type.tp_traverse(self, visit, arg);
+}
+
+static int
+abcmeta_clear(abc *tp)
+{
+    if (tp->abc_registry) {
+        PySet_Clear(tp->abc_registry);
+    }
+    if (tp->abc_cache) {
+        PySet_Clear(tp->abc_cache);
+    }
+    if (tp->abc_negative_cache) {
+        PySet_Clear(tp->abc_negative_cache);
+    }
+    return PyType_Type.tp_clear((PyObject *)tp);
+}
+
+static PyObject *
+abcmeta_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    abc *result = NULL;
+    PyObject *ns, *bases, *items, *abstracts, *is_abstract, *base_abstracts;
+    PyObject *key, *value, *item, *iter;
+    Py_ssize_t pos = 0;
+    result = (abc *)PyType_Type.tp_new(type, args, kwds);
+    if (!result) {
+        return NULL;
+    }
+    /* Set up inheritance registry. */
+    result->abc_registry = PySet_New(NULL);
+    result->abc_cache = PySet_New(NULL);
+    result->abc_negative_cache = PySet_New(NULL);
+    if (!result->abc_registry || !result->abc_cache || !result->abc_negative_cache) {
+        Py_DECREF(result);
+        return NULL;
+    }
+    result->abc_negative_cache_version = abc_invalidation_counter;
+    if (!(abstracts = PyFrozenSet_New(NULL))) {
+        Py_DECREF(result);
+        return NULL;
+    }
+    /* Compute set of abstract method names in two stages:
+       Stage 1: direct abstract methods.
+       (It is safe to assume everything is fine since type.__new__ succeeded.) */
+    ns = PyTuple_GET_ITEM(args, 2);
+    items = PyMapping_Items(ns); /* TODO: Fast path for exact dicts with PyDict_Next */
+    for (pos = 0; pos < PySequence_Size(items); pos++) {
+        item = PySequence_GetItem(items, pos);
+        key = PyTuple_GetItem(item, 0);
+        value = PyTuple_GetItem(item, 1);
+        is_abstract = _PyObject_GetAttrId(value, &PyId___isabstractmethod__);
+        if (!is_abstract) {
+            if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
+                Py_DECREF(item);
+                Py_DECREF(items);
+                goto error;
+            }
+            PyErr_Clear();
+            Py_DECREF(item);
+            continue;
+        }
+        if (is_abstract == Py_True && PySet_Add(abstracts, key) < 0) {
+            Py_DECREF(item);
+            Py_DECREF(items);
+            goto error;
+        }
+        Py_DECREF(item);
+    }
+    Py_DECREF(items);
+    /* Stage 2: inherited abstract methods. */
+    bases = PyTuple_GET_ITEM(args, 1);
+    for (pos = 0; pos < PyTuple_Size(bases); pos++) {
+        item = PyTuple_GET_ITEM(bases, pos);
+        base_abstracts = _PyObject_GetAttrId(item, &PyId___abstractmethods__);
+        if (!base_abstracts) {
+            if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
+                goto error;
+            }
+            PyErr_Clear();
+            continue;
+        }
+        if (!(iter = PyObject_GetIter(base_abstracts))) {
+            goto error;
+        }
+        while ((key = PyIter_Next(iter))) {
```

Comment:
Missing check for iterators that raise an error (such as when __abstractmethods__  is overwritten with a custom object).